### PR TITLE
Bugfix/concurrency

### PR DESCRIPTION
**Pull request by wmorgan6796**
The issue manifests itself at high throughput and high concurrency:
1.) When you go from read lock to write lock, there is enough of a pause where another thread can write to the aggregator, causing the payload to grow too large
2.) You didn't take into account the size of the partition_key_index when tracking the size of the payload.

These 2 issues break the library at high-concurrency and high throughput.

Also upgrading to go modules. Let me know if you want me to remove that.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,3 @@
 draft
+go.sum
```
Comment (a8m) [inline .gitignore]:
```suggestion
```

`go.sum` should be added as well.

File: .gitignore [hunk 2]
```diff
@@ -1 +1,3 @@
...
+*/.idea/
```
Comment (a8m) [inline .gitignore]:
```suggestion
```

You can add it to your local global .gitignore file, and avoid adding this to each project.

File: aggregator.go [hunk 3]
```diff
@@ -36,10 +35,12 @@ func (a *Aggregator) Put(data []byte, partitionKey string) {
 	// later, we will add shard-mapper same as the KPL use.
 	// see: https://github.com/a8m/kinesis-producer/issues/1
 	if len(a.pkeys) == 0 {
-		a.pkeys = append(a.pkeys, partitionKey)
+		a.pkeys = []string{partitionKey}
 		a.nbytes += len([]byte(partitionKey))
 	}
 	keyIndex := uint64(len(a.pkeys) - 1)
+
+	a.nbytes += 8
```
Comment (a8m) [inline aggregator.go]: Let's move this into a const at the top level declaration and use it in the producer.go as well.

**Comment by wmorgan6796** [comment]
Made the requested changes. Let me know if anything else stands out.

File: aggregator.go [hunk 4]
```diff
@@ -3,7 +3,6 @@ package producer
 import (
 	"bytes"
 	"crypto/md5"
-
```
Comment (a8m) [inline aggregator.go]: Add it back please.

**Pull request merged**